### PR TITLE
Introduce `DiplomatStr16`

### PR DESCRIPTION
**Pull request by robertbastian**
#57 

This adds the `DiplomatStr16` Rust type, and adds FFI bindings in JS (`String`), Dart (`String`), C2 (`wchar_t`), and CPP2 (`std::wstring_view`).

Based on #367

### CODE REVIEW

File: tool/src/cpp2/ty.rs [hunk 1]
```diff
@@ -537,8 +542,19 @@ impl<'ccx, 'tcx: 'ccx, 'header> TyGenContext<'ccx, 'tcx, 'header> {
                     expression: format!("{cpp_name}.AsFFI()").into(),
                 }]
             }
-            Type::Slice(hir::Slice::Str(..)) => {
-                // TODO: This needs to change if an abstraction other than std::string_view is used
+            Type::Slice(hir::Slice::Str(_, hir::StringEncoding::UnvalidatedUtf8)) => {
```
Comment (Manishearth) [inline tool/src/cpp2/ty.rs:545]: nit: can we just merge these two branches?

**Review by Manishearth — APPROVED** [review]

File: tool/src/cpp2/ty.rs (hunk 1 above)
Comment (robertbastian) [reply tool/src/cpp2/ty.rs]: Wildcard is fine, all our string types are and will be some kind of slice with a pointer and a length

File: tool/src/js/runtime.mjs [hunk 2]
```diff
@@ -1,15 +1,20 @@
-export function readString(wasm, ptr, len) {
+export function readString8(wasm, ptr, len) {
   const buf = new Uint8Array(wasm.memory.buffer, ptr, len);
   return (new TextDecoder("utf-8")).decode(buf)
 }
 
+export function readString16(wasm, ptr, len) {
+  const buf = new Uint16Array(wasm.memory.buffer, ptr, len);
+  return String.fromCharCode(buf)
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply tool/src/js/runtime.mjs:8]: I'm no JS guru, maybe there's a better way to make a `String` from a WASM ptr and length, but this works for now.

[the remark this answers is not in the document]
Comment (sffc) [reply tool/src/js/runtime.mjs:8]: This is close but not quite right. `String.fromCharCode` takes one or more individual arguments where each argument is a UTF-16 code unit.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCharCode

```
console.log(String.fromCharCode(189, 43, 190, 61));
// Expected output: "½+¾="
```

You need to use `.apply()` on the function in order to convert the Uint16Array to an argument list and get the behavior you want. There's an example here:

https://developer.chrome.com/blog/how-to-convert-arraybuffer-to-and-from-string/

For example:

```javascript
u16a = new Uint16Array([189, 43, 190, 61]);
String.fromCharCode.apply(null, u16a); // "½+¾="
```

(I think the code you wrote, without the `.apply()`, doesn't work at all; did you test it?)

File: tool/src/js/runtime.mjs [hunk 3]
```diff
@@ -78,8 +83,16 @@ export class DiplomatBuf {
...
+    })
+  }
+
+  static str16 = (wasm, string) => {
+    return new DiplomatBuf(wasm, string.length, 2, buf => {
+      for (var i; i < string.length; i++) {
+        buf[i] = string.codePointAt(i);
+      }
+    })
+  }
```
Comment (sffc) [inline tool/src/js/runtime.mjs:95]: I think you want `.charCodeAt`, not `.codePointAt`

**Review by sffc — COMMENTED** [review]
I saw someone mention JS string conversions so I looked at just that one file.

File: tool/src/js/runtime.mjs (hunk 2 above)
[the remark this answers is not in the document]
Comment (robertbastian) [reply tool/src/js/runtime.mjs:8]: Thanks, fixed in #371

**Comment by hsivonen** [comment]
> C2 (`wchar_t`), and CPP2 (`std::wstring_view`).

Why `wchar_t` and `std::wstring_view`? Neither should ever be used in portable code, since `wchar_t` on Windows has UTF-16 semantics and on most other present-day-relevant platforms has UTF-32 semantics.

#240 deliberately asked for `char16_t` and `std::u16string_view` as these have UTF-16 semantics regardless of Windows vs. other.

**Comment by robertbastian** [comment]
That's an oversight. I didn't know that `w` wasn't portable. Will fix.

**Comment by robertbastian** [comment]
#379

**Pull request merged**